proposals/0000-hasfield-redesign.rst

```diff
@@ -0,0 +1,1102 @@
+because there is no way to infer the field label from the record type or field
+type.
+
+If and when support for visible foralls is added, the type of ``getField`` could
+change to::
+
+  getField :: forall r a {k} . forall (x :: k) -> r -> a
+
+meaning that we could instead use ``getField "foo"`` at use sites.  (Per the
+visible forall proposal, here ``"foo"`` is a type-level ``Symbol`` even though
+it syntactically resembles a ``String`` literal.)
+
+This would be a breaking change, and visible dependent quantification is not yet
+fully implemented, so changing ``getField`` and ``setField`` to use it is not
+part of the present proposal.
+
+
+Pattern synonyms
+~~~~~~~~~~~~~~~~
+An infelicity with the current constraint solving behaviour for ``HasField`` is
+that it does not work for record pattern synonyms.  Thus where
+``OverloadedRecordDot`` or similar is used, replacing a datatype with an
+equivalent record pattern synonym may require declaring manual ``HasField`` and
+``SetField`` instances.
+
+It would be relatively easy to extend the automatic behaviour to support single
+record pattern synonyms.  For example, given the declaration::
+
+  pattern MyPair{car,cdr} = (car, cdr)
+
+it would be possible to solve a constraint like::
+
+  HasField "car" (x, y) x
+
+and hence a declaration like this would be accepted::
+
+  swap :: (x, y) -> (y, x)
+  swap p = MyPair { car = p.cdr, cdr = p.car }
+
+However, the fact that pattern synonyms can be added for arbitrary types (in
+this example, for the built-in type of pairs) mean that such behaviour can give
+rise to incoherent solutions to ``HasField`` constraints (cf. `proposal #515
+<https://github.com/ghc-proposals/ghc-proposals/pull/515>`_). For example, if
+another module defined::
+
+  pattern MyPair2{car,cdr} = (cdr, car)
+
+then the constraint ``HasField "car" (x, x) x`` would be solved differently
+depending on whether ``car`` from ``MyPair`` or from ``MyPair2`` was in scope.
+
+Moreover, it is unclear how to extend the automatic treatment of pattern
+synonyms to handle multiple-constructor types.  For example, given the
+declarations::
+
+  pattern MyLeft{val}  = Left val
+  pattern MyRight{val} = Right val
+
+we would ideally generate a solution to ``HasField "val" (Either a a) a`` that
+used both patterns, as in::
+
+  get_val :: Either a a -> a
+  get_val MyLeft{val} = val
+  get_val MyRight{val} = val
+
+However, it is not clear how to do this in general, since pattern synonyms are
+not necessarily grouped and may overlap in arbitrarily complex ways.  (While
+``COMPLETE`` pragmas do give a notion of grouping for pattern synonyms, their
+purpose is currently limited to the pattern-match completeness checker, and is
+not clear that they should have a semantic impact.)
+
+
+Unresolved Questions
+--------------------
+Changing ``SetField`` to support type-changing update is deliberately left out
+of this proposal, so that it can be considered in detail as a subsequent
+proposal.
```

Comment:
Is there a design for that already in someone's back's head? Is it going to be a nice extension of what we have here, or a major redesign? (I.e. is it really simpler to do that separately, of better to do all at once?)

Reply:
There is a half-finished (well, perhaps more like 1/4 finished) design here: https://github.com/adamgundry/ghc-proposals/blob/type-changing-update/proposals/0000-setfield-type-changing-update.rst. It is explicitly structured as an extension to this design (partly because it was originally part of the same draft proposal, before I cut it to make the whole thing more manageable).

The scope of this proposal is something I've struggled with. Type-changing update brings in a lot of complications and potential for things getting derailed, which is why I think it makes sense to try to establish consensus about the (hopefully) less controversial changes first. I don't expect that any of the changes proposed here will need to be significantly rethought in order to add type-changing update later. Some of them are somewhat motivated by it (e.g. splitting into two classes means that `HasField` can be left entirely alone by type-changing update) but I've tried to argue in this proposal that they are independently justifiable, even if type-changing update never materialises.

Reply:
Thanks!

Reply:
Thanks Adam, I think it's sensible to postpone tackling type-changing update. Somebody needing it can revert to using record syntax for pattern-matching on the incoming record, and explicit building the new record with the data constructor prefix. As they can now, so it's not taking anything away.

(I am curious why people are itching for special syntax for update -- as opposed perfectly workmanlike syntax for building a new record. I suppose I should say "so-called 'update'", because there's no _in situ_ update going on like imperative/OO languages: the program is going to build a new record either way; all we're talking about is syntactic sugar.)

IMO the H98 syntax for record update is horribly confusing. I never use it. I wish we could take this opportunity to completely redesign it.

Reply:
> IMO the H98 syntax for record update is horribly confusing. I never use it. I wish we could take this opportunity to completely redesign it.

My sense is that this ship has sailed, and that introducing a new syntax for updates now is not worth it (given that the existing one will remain for essentially forever). But I'm deliberately trying to steer away from such issues in this proposal, with `HasField` etc. functioning as a common substrate regardless of which syntax one uses.